### DISCUSSION TITLE
Does PikePDF support filling of XFA forms?

### DISCUSSION THREAD

**Discussion by rupeshbhurke** [pikepdf/pikepdf#526] [Q&A] [answered]
Hi!

I am working on a project that auto fills the the forms. The forms I have encoutered gets filled as long as I provide values for Text and Checkboxes type of filelds. The moment Radio button fields are involved. The values for filled fields may or may not show.

I have posed this question on other forums too. There I was informed that the PDF I am looking to fill is of type XFA.

I am attaching the PDF I am looking to fill.

Kindly advise.

Thank you.
[Primary_I-129.pdf](https://github.com/pikepdf/pikepdf/files/12789779/Primary_I-129.pdf)

**Comment by jbarlow83** [comment, marked as answer]
pikepdf can create and edit almost any valid PDF, including those with an XFA payload, and ensure that the output is a valid PDF. However, it does not have any code to specifically create XFA-conforming data payloads, or execute PDF Javascript like the Javascript present in the attached PDF, or process XFA in anyway. You would have to find some way to manage XFA, as pikepdf can only any payload is correctly inserted in the output PDF.

XFA is deprecated as PDF 2.0 - it's not a standard part of the ISO PDF specification, and it won't be around forever.

**Reply by rupeshbhurke** [reply]
Thank you for a quick reply. However I failed to find any API to fill form fields for PikePDF. 

kindly advise.

Thank you.

**Reply by jbarlow83** [reply]
`pdf.Root.AcroForm.XFA` is where XFA data is stored. You can download the PDF 2.0 specification and see Annex K. pikepdf will let retrieve and XFA data blobs (`.XRA_read_bytes()`), but cannot process them or fill them out.

If you are interested I can put you in touch with another user who I know was working on XFA using pikepdf.

**Reply by rupeshbhurke** [reply]
Thank you so much, But I have found another solution.

**Reply by sirpqp** [reply]
> Thank you so much, But I have found another solution.

hi,what's the another solution? recently i meet the problem too，it mess me up

**Reply by mehdiouchcham** [reply]
I would love to know the solution too, stuck on the same issu. Please advise. thank you.
